File: cses/05_range_queries/3304_visible-buildings-queries/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    pos = 0
    n = int(data[pos]); pos += 1
    q = int(data[pos]); pos += 1

    # Chiều cao các tòa nhà, đánh chỉ số từ 1.
    h = [0] * (n + 2)
    for i in range(1, n + 1):
        h[i] = int(data[pos]); pos += 1

    INF = n + 1  # sentinel: luôn > b vì b <= n

    # nge[i] = tòa đầu tiên bên phải cao hơn NGHIÊM NGẶT tòa i, else INF.
    # Tính bằng monotonic stack O(n), duyệt từ phải sang trái.
    nge = [INF] * (n + 2)
    stack = []
    for i in range(n, 0, -1):
        hi = h[i]
        # Loại các tòa không cao hơn tòa hiện tại khỏi đỉnh stack.
        while stack and h[stack[-1]] <= hi:
            stack.pop()
        nge[i] = stack[-1] if stack else INF
        stack.append(i)
    nge[INF] = INF  # trạng thái hấp thụ cho sentinel

    # Bảng binary lifting: up[k][i] = vị trí sau khi nhảy 2^k bước nge từ i.
    LOG = 1
    while (1 << LOG) < n + 1:
        LOG += 1
    up = [nge]
    for k in range(1, LOG):
        prev = up[k - 1]
        cur = [0] * (n + 2)
        for i in range(1, n + 2):
            cur[i] = prev[prev[i]]
        up.append(cur)

    # Trả lời truy vấn: đếm số bước nhảy còn nằm trong [a, b].
    out = []
    ap = out.append
    for _ in range(q):
        a = int(data[pos]); pos += 1
        b = int(data[pos]); pos += 1
        node = a
        ans = 1  # tòa a luôn visible
        # Duyệt tầng từ cao xuống thấp, nhảy khi vẫn còn trong đoạn.
        for k in range(LOG - 1, -1, -1):
            nxt = up[k][node]
            if nxt <= b:
                node = nxt
                ans += 1 << k
        ap(ans)

    sys.stdout.buffer.write(("\n".join(map(str, out)) + "\n").encode())


main()

File: cses/05_range_queries/3304_visible-buildings-queries/solution.cpp
#include <iostream>
#include <vector>

int main() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);

    int n, q;
    std::cin >> n >> q;

    // Chiều cao các tòa nhà, đánh chỉ số từ 1.
    std::vector<int> height(n + 2, 0);
    for (int building = 1; building <= n; ++building) {
        std::cin >> height[building];
    }

    // next_greater[i] = tòa đầu tiên bên phải cao hơn NGHIÊM NGẶT tòa i,
    // nếu không có thì bằng sentinel = n+1. Tính bằng monotonic stack O(n).
    const int sentinel = n + 1;
    std::vector<int> next_greater(n + 2, sentinel);
    std::vector<int> stack;
    stack.reserve(n);
    for (int building = n; building >= 1; --building) {
        // Loại các tòa không cao hơn tòa hiện tại khỏi đỉnh stack.
        while (!stack.empty() && height[stack.back()] <= height[building]) {
            stack.pop_back();
        }
        if (!stack.empty()) {
            next_greater[building] = stack.back();
        }
        stack.push_back(building);
    }
    next_greater[sentinel] = sentinel;  // trạng thái hấp thụ cho sentinel

    // Số tầng binary lifting: LOG ~ ceil(log2(n)).
    int levels = 1;
    while ((1 << levels) < n + 1) {
        ++levels;
    }

    // jump[level][i] = vị trí sau khi nhảy 2^level bước theo con trỏ next_greater.
    std::vector<std::vector<int>> jump(
        levels, std::vector<int>(n + 2, sentinel)
    );
    jump[0] = next_greater;
    for (int level = 1; level < levels; ++level) {
        for (int building = 1; building <= sentinel; ++building) {
            jump[level][building] =
                jump[level - 1][jump[level - 1][building]];
        }
    }

    // Trả lời truy vấn: đếm số bước nhảy còn nằm trong [first, last].
    while (q-- > 0) {
        int first, last;
        std::cin >> first >> last;
        int visible = 1;          // tòa first luôn visible
        int building = first;
        // Duyệt tầng từ cao xuống thấp, nhảy khi vẫn còn trong đoạn.
        for (int level = levels - 1; level >= 0; --level) {
            if (jump[level][building] <= last) {
                building = jump[level][building];
                visible += 1 << level;
            }
        }
        std::cout << visible << '\n';
    }
    return 0;
}
